File: leetCode/fukun/101-200/lc_142.js
var detectCycle = function (head) {
    let currs = head, currf = head, meet = null;
    while (1) {
        if (!currf || !currf.next) return null;
        currf = meet ? currf : currf.next;
        [currs, currf] = [currs.next, currf.next];
        if (currs === currf) {
            if (currs === head) return head;
            if (!meet) {
                [meet, currs] = [currs, head];
            } else {
                return currs;
            }
        }
    }
};

File: leetCode/linovce/lc_142.java

  Definition for singly-linked list.
  class ListNode {
      int val;
      ListNode next;
      ListNode(int x) {
          val = x;
          next = null;
      }
  }
 
public class Solution {
    public ListNode detectCycle(ListNode head) {
        ListNode q = head;
        ListNode p = head;
        
        while(q!=null){
            q=q.next;
            if(q!=null)
                q=q.next;
            else
                return null;
            p=p.next;
            
            if(q==p){
                q=head;
                while(q!=p){
                    q=q.next;
                    p=p.next;
                }
                return q;
            }
        }
        return null;
    }
}